### PR TITLE
feat: Sound effect text for bows, crossbows, launchers, railguns and pneumatic weapons

### PR DESCRIPTION
**Pull request by Soadreqm**
### Required

- [x] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [x] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [x] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [x] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

## Purpose of change
After #5309, some weapons that were supposed to have their own special sound effect text were instead printing generic gun noises. While fixing this, I ended up adding new effect text for different sound levels for railguns and pneumatic weapons. Some additional little oddities with sound levels were discovered, and an attempt was made to fix these.

## Describe the solution
There's now a bunch of new item flags for what sound a weapon is supposed to make, which are checked when writing the sound description text.

Grenade launchers were supposed to have their sound level hardcoded to 8. This didn't work, which is just as well since the grenade detonation doesn't seem to make a sound of its own. New behaviour is still broken: launcher grenades make a large boom, but the sound still originates from the player, not the grenade detonation point. 

Railguns were supposed to have their sound level hardcoded to 24. This worked for the H&K G80, but not for the homemade rail rifles. It seems a bit odd anyway, since a large hypersonic projectile would probably be pretty loud. The special hardcoding was removed, and the weapons were given explicit json loudness values: the G80 can stay at 24 with its pre-cataclysm teutonic gnome magic, but the homemade rail rifles were set to 180 and 220 for the light and heavy version. The coilgun was set to 25 loudness, and the basic nail has 5 more loudness over the flechette due to poor aerodynamics. And since I changed the nail default noise value, I also gave some for other nailguns.

Flamethrowers and rocket launchers were supposed to have their sound level hardcoded to 4. Did not change, since it seems to actually work as intended - firing the rocket is quiet, and the detonation is what makes all the noise.

The pneumatic air pump attachment adds 16 noise now. I think I meant to put that in a separate PR at some point, but I was testing sound levels with it.

## Describe alternatives you've considered
Thought about adding sound effect markers to ammo instead of (or in addition to) guns. Or adding a new json field for it, instead of a flag. This seemed easiest, but I'm honestly not sure if it's the best approach. It works but it's not particularly extensible. I don't know anything about sound packs, so I have no idea if any of this is going to be useful with those.

## Testing
Spawned and fired a large number of different guns, launchers and bows. Checked that they printed the right sound effect, and that the sound value looked right.

### CODE REVIEW

File: src/ranged.cpp [hunk 1]
```diff
@@ -1851,21 +1841,38 @@ item::sound_data item::gun_noise( const bool burst ) const
 
     noise = std::max( noise, 0 );
 
-    if( ammo_current() == itype_40x46mm || ammo_current() == itype_40x53mm ) {
+    if( has_flag( flag_SOUND_LAUNCHER ) ) {
         // Grenade launchers
-        return { 8, _( "Thunk!" ) };
+        return { noise, _( "Thunk!  Boom!" ) };
```
Comment (chaosvolt) [inline src/ranged.cpp:1846]: Side note, not all grenade launcher ammo will explode, e.g. the buckshot rounds.

Comment (chaosvolt) [inline src/ranged.cpp:1846]: Given discussion and testing in the discord, we can try and fix frag ammo not making any noise in a later PR I figure, so this probably might work well enough for now. Also went with something a lil more different than the crossbow noise, if desired.
```suggestion
        return { noise, _( "Thump!" ) };
```

**Review by scarf005 — DISMISSED** [review]
concerns:

1. I think default weapon firing sound should follow weapon categories. This approach causes more code duplications.
2. Flags are also less flexible by default, having only 6 variants in this case. It might be better to add property like `"gun_sound"` so mod developers could customize how it sounds.

asking @Coolthulhu for input.

**Comment by Soadreqm** [comment]
> 2\. It might be better to add property like `"gun_sound"` so mod developers could customize how it sounds.

Do you mean just `"gun_sound": "RAILGUN"`, or moving the sound effect string completely to json? I don't think the latter would be a good idea. It'd make it impossible to adjust the sound to gun volume, if it's modded to be louder or quieter, unless you also included the noise thresholds. You'd be looking at something like
`"gun_sound": [ [ "plink", 0 ], [ "bang!", 10 ], [ "blam!", 150 ], [ "kerblam!", 175 ] ],`
`"gun_sound_burst": [ [ "Brrrip!", 0 ], [ "Brrrap!", 10 ], [ "P-p-p-pow!", 150 ], [ "Kaboom!", 175 ] ]`
to replicate current behaviour with normal guns. And of course if you ever wanted to change the effects or level thresholds for whatever reason, you'd have to do it item by item, at least for anything that doesn't `"copy-from": "gun_base"`.

**Comment by chaosvolt** [comment]
I started tinkering with the idea last night. My idea was to define a property for the sound string of gun firing and for bursts like you proposed, but leave in the fallback behavior for normal gun sounds and have it only print gun sounds if defined, like this:
```C++
    // If firing noise defined, use it. If we can use a separate noise for bursts then do so, otherwise general firing noise.
    if( !type->gun->firing_noise_burst.empty() && burst ) {
        return { noise, type->gun->firing_noise_burst };
    }
    if( !type->gun->firing_noise.empty() ) {
        return { noise, type->gun->firing_noise };
    }
```

However, you're right in that this still has the flaw of no longer varying by noise level. Leaving the default "gun go bang" handling in reduces the problem this causes somewhat, so of the existing volume-dependent gun noises only laser/plasma and lightning weapons would be affected by this loss (though we would also lose the railgun and pneumatic sound-by-volume variation this PR introduces).

This could however be compensated for manually at the level of whichever guns in those categories have default loudness noticably higher or lower than what the majority use, but this would require being hashed out and tested for manually. It's not something I'd want to mandate another user bother with, but I could do so myself if you'd want me to. If so it'll likely have to wait until later today since going to be out and about in a bit.

**Comment by Coolthulhu** [comment]
@scarf005 

> I think default weapon firing sound should follow weapon categories. 

I think flags are better, but I'm not sure about it. Flags are more explicit and can be forgotten, but categories are too broad.

@Soadreqm 
> Do you mean just "gun_sound": "RAILGUN", or moving the sound effect string completely to json?

Ideally both:

* On a gun, `"gun_sound": "RAILGUN"`
* In other file:
```json
{ "type": "gun_sound", "id": "RAILGUN", "sounds": [ [ 10, "pew pew pew!" ], [ 100, "WOOM!" ] ] }
```

**Comment by Soadreqm** [comment]
#5339 does the same thing, and leaves ranged.cpp looking prettier, so I'm gonna close this one. There's a couple ancillary changes here, like tweaks to nailgun noise, but those are better done separately anyway. I'll do a new PR for them later, if Chaosvolt doesn't beat me to it.

**Comment by chaosvolt** [comment]
No worries. And yeah, weapon noises will need an audit eventually. I think nailguns and railguns seem okay more or less, the bigger issues tend to be stuff like:
1. Weapons that're totally silent but maybe shouldn't be (I noticed the slingshot cannon was a double example of this, it both shouldn't be in terms of I feel it should make some noise, and it shouldn't be in terms of it's reporting its firing sound as zero despite the gun having some loudness).
2. Launcher weapons that're deafening due to HEAT/HEDP shenanigans that should merely be loud but not eardrum-destroying (most notably 40mm guns).

Related to point two is how frag ammo effects don't generate any noise when they explode, something I need to look at.

**Comment by chaosvolt** [comment]
Also lemme know if I missed anything else in this PR that should go in the other one instead of being saved for the audit. ^^

**Pull request closed without merging**